### ISSUE TITLE
Git commit crashing with mirror.exe

### ISSUE THREAD

**Issue by dimov-cz** [dokan-dev/dokany#369]
# Environment
 * Windows version: 10x
 * Processor architecture: x64
 * Dokany version: 1.0.0.5
 * Library type (Dokany/FUSE): Dokany
 
# Check List
  - [x] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [x] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [x] I can always reproduce the issue with the provided description below.
  - [x] I have updated Dokany to the latest version (https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.

# Description
I'm little bit dissappointed from 1.0.0.5, previous version(RC4) was working better for me. A lot of people now reports me in WinSshFS about not working git. 
I can confirm this with reproduction script for mirror.exe. 

**Before running this test i must warn you, you will very probably get explorer.exe crash, you will be unable to logout, open start menu, processes can go to some uninterruptible state and hardreset will be only option probably.**

1. Create mirror: `mirror.exe /r mirrored /l t /t 1 /d /s`
2. git clone some small repo and try to commit.
I use this test script: [wfsGitTest1.zip](https://github.com/dokan-dev/dokany/files/525713/wfsGitTest1.zip) 

If you restart mirror for second try u are unable to remove directory over mirror.
Maybe it's related to something already fixed, but not released. This one? https://github.com/dokan-dev/dokany/issues/345

**Comment by Liryna** [comment]
Hi @dimov-cz ,

Yes the issue is related to #345 . I tested your script on a snapshot and it work 👍 
```
liryna@DESKTOP-T9B1IFA MINGW64 ~/Desktop
$ ./wfsGitTest1.sh
Cloning into 'emsdk.tmp'...
remote: Counting objects: 825, done.
remote: Compressing objects: 100% (8/8), done.
remote: Total 825 (delta 2), reused 0 (delta 0), pack-reused 817
Receiving objects: 100% (825/825), 393.60 KiB | 0 bytes/s, done.
Resolving deltas: 100% (517/517), done.
warning: LF will be replaced by CRLF in newfile.dat.
The file will have its original line endings in your working directory.
[master 0bcd92e] add new file
 1 file changed, 1 insertion(+)
 create mode 100644 newfile.dat
0
```

Sorry for the issues in 1.0.0, the lock unlock FCB has been changed after RC4 and it has not been enough tested. A next 1.0.1 will come soon with all the fixes ! 😢

**Comment by dimov-cz** [comment]
Sorry, but not for me, with latest snap. Cloning is ok. But commit fails. The error is different on current master (3babcb): fatal: mmap failed: Invalid argument

**Comment by Kerbox** [comment]
At least it isn't crashing anymore. I note that if you try to commit one more time, it works.

The main difference I can see in procmon is that for NTFS, CreateFileMapping returns "FILE LOCKED WITH ONLY READERS", while in mirror, it returns "FILE LOCKED WITH WRITERS".
Another difference is that when in mirror, we get a FileSystemControl operation, that I cannot see when using NTFS. So that is strange.

**Comment by Liryna** [comment]
Oh sorry @dimov-cz , I tested again and it is right that there is a mmap error.
@Kerbox is right it is the second commit that work strangely. When I tested it, first commit proposed me to config my git environment so I had no error 😢

**Comment by Liryna** [comment]
This work with 1.0.0-RC4.

I debugged the current master and it seems like there is some unlock not proceed here.
I have not been able to find the exact one but I saw a couple of Lock made not followed by an Unlock 😢 

@taruti do you think you can help on this ? I feel that some scope of locks (specially from create.c) are too big and could be reduced not ?

**Comment by taruti** [comment]
@Liryna I'll look at this on thu.

**Comment by Kerbox** [comment]
So, I've been researching this problem a bit. I saw that after cloning the repository, a CreateFileMapping of .git\index fails, due to GetFileInfo with invalid handle.

It looks like a Cleanup has been done on the .git\index during the clone process, but never a Close. Once you try to do a CreateFileMapping, you get a GetFileInfo and then a Close on the previous handle!

If I remove
```
    CloseHandle((HANDLE)(DokanFileInfo->Context));
    DokanFileInfo->Context = 0;
```
from MirrorCleanup, git clone and commit works again.

**Comment by taruti** [comment]
Just to note that `git commit` appears to work with a non-mirror filesystem.

**Comment by Liryna** [comment]
I see ! So probably we should also open the file in GetFileInfo when the handle is invalid (already called after cleanup).

@taruti I tought there is/was a lock issue because after running the script, leave the folder/explorer close the mirror cleanly. The kernel device does not want to be unmount because there is still reference on files looks like.

**Comment by Kerbox** [comment]
Yes, I can fix it so it works like Read and Write.

**Comment by Kerbox** [comment]
I cannot get it to work, if I do it the same way as in Read, I get "bad index file sha1 signature" from git. 
I have no idea why, and I do not have any more time this week to find out unfortunately.
Also, the directory structure that git creates during clone, cannot be deleted through explorer for some reason.

**Comment by taruti** [comment]
`$ ./wfsGitTest1.sh` also seems to work for me. 1.0.1.0795.

**Comment by Liryna** [comment]
@taruti 1.0.1.0795 ? so you mean it work before https://github.com/dokan-dev/dokany/commit/63655e3410ba8982669018deff48004f728a2b36 ?

**Comment by taruti** [comment]
@Liryna, haven't done a proper bisect yet.

**Comment by Liryna** [comment]
@taruti I have start to review the FCB lock (to reduce the scope of it and try to not have oplock calls in the lock)
https://github.com/dokan-dev/dokany/tree/review-fcb-lock

I remember you reviewed it because of BSOD when copy a lots of files. Did you just copy something like 1000 files and get a bsod ? (just to be sure that I dont bring back the bsod)

**Comment by taruti** [comment]
I had a test case with a filetree of ~66k small files and directories that I copied over and deleted.

**Comment by Liryna** [comment]
@Kerbox I also get "bad index file sha1 signature" when I tried to change it 😢 

I added support of "FILE LOCKED WITH ONLY READERS". It now correctly return the value but it does not change the behavior 😢 
https://github.com/dokan-dev/dokany/commit/2768ac75402a0dbe06de163763485e251be2ac54

I don't know what is still wrong for him and why there is still reference on the device after running the script, it makes the device unable to be removed on the kernel side 😞 https://github.com/dokan-dev/dokany/blob/da39829ddccbf68e7d4af3966edb8b0c91afa25f/sys/init.c#L498

For information, clone an empty repository also reproduces the issue (it helps for having a cleaner debug output).

Ok @taruti ! I will try to do something like that !
Your changes are perfect! I just forgot that oplock functions use callbacks that does not clean the lock sometimes 😢

**Comment by taruti** [comment]
I'll do a PR with less locking in a few hours if you'd like :)

**Comment by Liryna** [comment]
@taruti I would ❤️ that !

**Comment by taruti** [comment]
https://github.com/dokan-dev/dokany/pull/382 but will test things with time on monday.

**Comment by Kerbox** [comment]
@Liryna If you haven't fixed this issue with the lock changes, I can resume trying to find out the reason tomorrow!

**Comment by Liryna** [comment]
This would be awesome @Kerbox ! because I found nothing for now 😢

**Comment by taruti** [comment]
@Kerbox btw does #382 help for you?

**Comment by Kerbox** [comment]
@taruti No, does not seem to make any difference.

**Comment by Liryna** [comment]
@Kerbox have you been able to take a last look at it ?
I will probably release the 1.0.1 without fixing this issue.

**Comment by Kerbox** [comment]
No luck yet, unfortunately. I cannot figure out why I get Close on NTFS, but not on mirror. 
I have a few other ideas I will try out later.

**Comment by Liryna** [comment]
Ok @Kerbox no problem, for information I have revert the previous commit that was creating a bsod
See https://github.com/dokan-dev/dokany/issues/386 I will put it back when https://github.com/dokan-dev/dokany/issues/387 will be done

**Comment by Kerbox** [comment]
I found out that the reason for the "bad index file sha1 signature" is that when renaming index.lock to index, there are two index FCB's, for the old and new. If I "fix" this, I can get git to work. But I suspect this is just masking the symptom, because I still have problem with the device references.

**Comment by Liryna** [comment]
Rename keep the old and new FCB 😮 ?
Does a simple rename file should not procude the same issue ?

I have add the  `return STATUS_FILE_LOCKED_WITH_ONLY_READERS;` as before https://github.com/dokan-dev/dokany/commit/2e2fea7e23c62e01fb2f7e369fb6d921a41a3ca8
and added the open handle when GetFileInformation requested after cleanup  https://github.com/dokan-dev/dokany/commit/7bdaf57818cb4086f08d9b240c8aa433b6c1da3d

**Comment by Kerbox** [comment]
@Liryna Yes, because .git\index have not received an IRP_MJ_CLOSE, when .git\index.lock get an IRP_MJ_SET_INFORMATION to rename to .git\index, it sets the FileName to .git\index. Thus, we have two FCB with .git\index. When a new IRP_MJ_CREATE with .git\index is received, it uses the old FCB!

I don't know specifically why we don't get IRP_MJ_CLOSE when using git, but I have reproduced the issue doing this (pseudocode):
```
filehandle = CreateFile()
maphandle = CreateFileMapping(filehandle)
mmap = MapViewOfFileEx(maphandle)
CloseHandle(maphandle)
```
If I now return from the program without unmapping the views or closing the filehandle, I never get IRP_MJ_CLOSE on mirror. But I do on NTFS.

**Comment by Liryna** [comment]
Oh ! If this can be reproduced with this pseudo-code, it gonna be simpler to debug it 😄 !

So it mean that we have a leak when there is a CloseHandle without the UnmapViewOfFile ?
There is an old discution about it https://github.com/dokan-dev/dokany/issues/32#issuecomment-128708655

Do you think you can give a full code test that I can test on my side with you ?

**Comment by Kerbox** [comment]
Can give you some test code later. I also have a few ideas I want to try out.

**Comment by Liryna** [comment]
Ok perfect @Kerbox !
Thanks again for all your help ❤️ !

**Comment by Kerbox** [comment]
@Liryna I probably won't be able to do much before the weekend, so don't hold up the release of 1.0.1 waiting for me!

**Comment by Liryna** [comment]
OK I will do it therefore 👍

**Comment by Kerbox** [comment]
7bdaf57 may fix it for @dimov-cz, but if he also still have the "bad index file" on winsshfs, a fix that will work is to return error from MoveFile if there are existing handles for the target file that has not been Closed.

**Comment by Liryna** [comment]
Hi @Kerbox,

Do you still have some work pending about this ?

**Comment by marinkobabic** [comment]
@Liryna Now we are doing flush which does close of memory mapped files. Does this issue still exist?

**Comment by Liryna** [comment]
@marinkobabic I tought about it too 👍  will try to test again later.
But currently have a hardware crash of my personnal computer 😢

**Comment by Kerbox** [comment]
This issue seems to have been fixed by #454 👍

**Comment by Liryna** [comment]
Hi  @Kerbox ,

You right, this was fixed with #454 but has since broked with https://github.com/dokan-dev/dokany/pull/462 😢 
@marinkobabic any idea what could happen ?

**Comment by marinkobabic** [comment]
The main difference is that in second pull request only part of the files
is going to be flushed. Log would provide more information. You can change
the code of the second request so that the filter is not taking place. This
should solve the problem. Then we can talk what is the source of the
problem.

Am 19.03.2017 09:34 schrieb "Liryna" <notifications@github.com>:

**Comment by Liryna** [comment]
Hi @marinkobabic ,

I test previous commit of #462 and it still work.
I have try to compare/revert, I have been able to get everything still work with PR #462 if I keep the old FlushAllCachedFcb.
I could not look more far, I tried to get some logs but found not calls of FlushAllCachedFcb at all.
Maybe one of the calls of FlushAllCachedFcb  is made with a null fcb and is skipped ? https://github.com/dokan-dev/dokany/pull/462/files#diff-34a56c4e800894f7ec40a0aaae302548R375

**Comment by Liryna** [comment]
@marinkobabic I have been able to look make FlushAllCachedFcb "work" again by removing the fcbRelatedTo folder check at the start and the StartsWith see patch: [0001-test.txt](https://github.com/dokan-dev/dokany/files/853354/0001-test.txt)
I know that with this patch it look like your preview PR where everything was flush...but that was the only way I was making it work.
I have run dbgview with the patch : [dbgview.txt](https://github.com/dokan-dev/dokany/files/853357/dbgview.txt)

There is not trace of FlushAllCachedFcb print, I dont really understand why 😢

**Comment by marinkobabic** [comment]
Please try the following if fileRelatedTo is null then

If fileObject is not null then we should execute FlushAllCachedFcb(fcb,
fileObject);

Am 19.03.2017 17:38 schrieb "Liryna" <notifications@github.com>:

**Comment by Liryna** [comment]
Humm @marinkobabic don't real get.
```	
if (fileObject != NULL)
   FlushAllCachedFcb(fcb, fileObject);
```	
Has to be added where ? (since there is fcb, I tought about inside the loop but we would not reach it with `if (fileRelatedTo) return` from the start.

**Comment by marinkobabic** [comment]
I meant FlushFcb(fcb, NULL);should be executed.

Am 19.03.2017 17:59 schrieb "Marinko Babic" <marinko.android@gmail.com>:

**Comment by Liryna** [comment]
`FlushFcb(fcb, NULL);` but what is the value of fcb ? if fileRelatedTo  is null like you said, there is no fcb
Could you be more precise 😃 I don't really see here ❤️

**Comment by marinkobabic** [comment]
When you apply your cache. What files are flushed.

DDbgPrint("  FlushAllCachedFcb check %wZ if is related to %wZ",
DDbgPrint("  FlushAllCachedFcb flush %wZ if flush is possible."

compared when your patch is not applied? You can set the filter in debugview to *FlushAllCachedFcb* so we would get the right output. Just wondering what the difference is and what we are missing.

**Comment by Liryna** [comment]
Here is 3 logs with `*FlushAllCachedFcb*` 👍 


Without my patch: [without-patch.txt](https://github.com/dokan-dev/dokany/files/853496/without-patch.txt)
With my patch: [with-patch.txt](https://github.com/dokan-dev/dokany/files/853494/with-patch.txt)
Without the first check if fcbRelatedTo is a directory: [with-startwith.txt](https://github.com/dokan-dev/dokany/files/853495/with-startwith.txt)

**Comment by marinkobabic** [comment]
StartsWith(**&fcb->FileName**, **&fcbRelatedTo->FileName**)) should be
StartsWith(**&fcbRelatedTo->FileName**, **&fcb->FileName**)

Does this change fix the issue already?

**Comment by Liryna** [comment]
Yes this fix the issue 👍 ! [fix-startswith.txt](https://github.com/dokan-dev/dokany/files/853557/fix-startswith.txt)
But I tried to renable the if fcbRelatedTo is a not directory and it failed: [fix-starwith-keep-isnotdirect.txt](https://github.com/dokan-dev/dokany/files/853559/fix-starwith-keep-isnotdirect.txt)

**Comment by marinkobabic** [comment]
Please add to your log the filters MmFlushImageSection and CcFlushCache

I'm even not sure about the change proposed above. If somebody renames \liryna\Desktop\emsdk.tmp\.git\config.lock	there is no reason to flush \liryna\Desktop\emsdk.tmp\.git\config

please change also this line DDbgPrint("  FlushAllCachedFcb file passed in. Flush only this file %wZ.\n", &fcbRelatedTo->FileName);

The original idea was if we have a directory \liryna\Desktop\mydir and if somebody renames this directory then we should flush all the files which start with \liryna\Desktop\mydir before the rename happens. Let's see what the result of log is.

**Comment by Liryna** [comment]
Log added 👍 `*FlushAllCachedFcb*;*MmFlushImageSection*;*CcFlushCache*`
Simple log without changes: [without-dir-check.txt](https://github.com/dokan-dev/dokany/files/853987/without-dir-check.txt)
Without the first if file check: [without-patch.txt](https://github.com/dokan-dev/dokany/files/853986/without-patch.txt)

**Comment by Liryna** [comment]
On the full logs, we can see that emsdk.tmp.git\config.lock is renamed to emsdk.tmp.git\config
```
00526399	11.05768108	[DokanFS] ==> DokanCompleteSetInformation	
00526400	11.05768394	[DokanFS]    rename also done on fcb \liryna\Desktop\emsdk.tmp\.git\config 	
00526401	11.05768681	[DokanFS]   DokanCompleteSetInformation Report FileRenameInformation
[DokanFS] ==> DokanNotifyReportChange \liryna\Desktop\emsdk.tmp\.git\config.lock	
00526402	11.05768871	[DokanFS] <== DokanNotifyReportChange	
00526403	11.05769062	[DokanFS] ==> DokanNotifyReportChange \liryna\Desktop\emsdk.tmp\.git\config	
00526404	11.05769157	[DokanFS] <== DokanNotifyReportChange	
00526405	11.05769253	[DokanFS]   status = 0x0	
00526406	11.05769444	[DokanFS]   status = STATUS_SUCCESS	
00526407	11.05769539	[DokanFS] <== DokanCompleteSetInformation	
```

**Comment by marinkobabic** [comment]
This helps a lot. Thanks. Will analyze the stuff in the evening.

Am 20.03.2017 07:48 schrieb "Liryna" <notifications@github.com>:

**Comment by marinkobabic** [comment]
@Liryna Please can you check if this simple fix solves the problem https://github.com/marinkobabic/dokany/commit/5d1a00edc5166063ea2d5d9162beadbab3a102fc

```
      targetFileObject = irpSp->Parameters.SetFile.FileObject;
      if (targetFileObject)
      {
          DDbgPrint("  FileRenameInformation targetFileObject specified so perform flush\n");
          PDokanCCB targetCcb = (PDokanCCB)targetFileObject->FsContext2;
          ASSERT(targetCcb != NULL);
          FlushAllCachedFcb(targetCcb->Fcb, targetFileObject);
      }
      FlushAllCachedFcb(fcb, fileObject);
```

**Comment by Liryna** [comment]
Hi @marinkobabic ,

Yes this work perfectly 👍 thank you !
You can create a PR with the commit and will merged it 😃 !

**Comment by Liryna** [comment]
Fixed with https://github.com/dokan-dev/dokany/pull/482 PR ❤️ !

**Comment by dimov-cz** [comment]
👍  Great news, thanks to everyone!
